src/qibo/derivative.py

```diff
@@ -0,0 +1,125 @@
+            import numpy as np
+            from qibo import hamiltonians, gates
+            from qibo.models import Circuit
+
+            # in order to see the difference with tf gradients
+            import tensorflow as tf
+            qibo.set_backend('tensorflow')
+
+            # defining an observable
+            def hamiltonian(nqubits = 1):
+                m0 = (1/nqubits)*hamiltonians.Z(nqubits).matrix
+                ham = hamiltonians.Hamiltonian(nqubits, m0)
+                return ham
+
+            # defining a dummy circuit
+            def circuit(nqubits = 1):
+                c = Circuit(nqubits = 1)
+                c.add(gates.RY(q = 0, theta = 0))
+                c.add(gates.RX(q = 0, theta = 0))
+                c.add(gates.M(0))
+                return c
+
+            # using GradientTape to benchmark
+            def gradient_tape(params):
+                params = tf.Variable(params)
+
+                with tf.GradientTape() as tape:
+                    c = circuit(nqubits = 1)
+                    c.set_parameters(params)
+                    h = hamiltonian()
+                    expected_value = h.expectation(c.execute().state())
+
+                grads = tape.gradient(expected_value, [params])
+                return grads
+
+            # initializing the circuit
+            c = circuit(nqubits = 1)
+
+            # some parameters
+            test_params = np.random.randn(2)
+            c.set_parameters(test_params)
+
+            # a test hamiltonian
+            test_hamiltonian = hamiltonian()
+
+            # running the psr with respect to the two parameters
+            grad_0 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 0, gate_eigenv = 0.5)
+            grad_1 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 1, gate_eigenv = 0.5)
+
+            # evaluating the gradients using tensorflow
+            tf_grads = gradient_tape(test_params)
+
+            print('Test gradient with respect params[0] with PSR: ', grad_0.numpy())
+            print('Test gradient with respect params[0] with tf:  ', tf_grads[0][0].numpy())
+            print('Test gradient with respect params[1] with PSR: ', grad_1.numpy())
+            print('Test gradient with respect params[1] with tf:  ', tf_grads[0][1].numpy())
+
+    """
+
+    if parameter_index > len(circuit.get_parameters()):
+        raise ValueError(
+            """This index doesn't exist! Use one of the circuit's parameters."""
+        )
+
+    # defining the shift according to the psr
+    s = np.pi / (4 * gate_eigenv)
+
+    # saving original parameters and making a copy
+    original = np.asarray(circuit.get_parameters()).copy()
+    shifted = original.copy()
+
+    # forward shift and evaluation
+    shifted[parameter_index] += s
+    circuit.set_parameters(shifted)
+
+    if initial_state is None:
+        forward = observable.expectation(circuit.execute().state())
```

Comment:
In the core qibo code I avoid using the `GlobalBackend` because it makes testing harder and also global objects are generally dangerous. Here `circuit.execute()` will fall back to the `GlobalBackend` so if someone defines the `observable` Hamiltonian with a different backend we may have issues, particularly if cupy is mixed with numpy.

Since `observable` is a `Hamiltonian` object, it already has an underlying backend, so the easiest would be to use this backend for circuit execution. This can be done with:
```suggestion
        backend = observable.backend
        forward = observable.expectation(backend.execute_circuit(circuit).state())
```
or you may prefer to define the `backend` variable in the beginning of the function since it will be reused multiple times.

Reply:
Before doing this, I want to be sure to have understood. 
How do I import the generic backend object?

Reply:
After the last large refactoring, there are two ways to execute circuits: one is to explicitly define a backend object and use it to execute a circuit:
```py
from qibolab.backends import NumpyBackend

circuit = Circuit(...)
backend = NumpyBackend() # can be replaced by TensorflowBackend
# or NumbaBackend() / CupyBackend() if you have qibojit installed
result = backend.execute_circuit(circuit)
```
With this approach is clear what backend you are using. The other approach is to just execute the circuit directly:
```py
circuit = Circuit(...)
result = circuit()
```
This will [automatically instantiate](https://github.com/qiboteam/qibo/blob/8b0e0d49cbab6eae7d550ae55081af9c0c0f448e/src/qibo/models/circuit.py#L994)  a `GlobalBackend` object which is a reference to one of the available backends (numpy, tensorflow, etc.). `qibo.set_backend` controls to which actual backend `GlobalBackend` is refering to. If you do not use `set_backend`, qibo will automatically fall back to the first available backend from the following list: cupy, numba, tensorflow, numpy (the first available in your environment).

The idea is that in the core code and tests we use the first approach because it is more explicit, know exactly which backend you are using. We just provide the `result = circuit()` approach to simplify circuit execution for users who do not wish to go through the backend details and also keep compatibility with older qibo versions where backends were handled differently. 

Note that the `GlobalBackend` is used internally to make this work, the user is not supposed to import it or use it directly. If they wish to get exposed to the different backends, they should import them directly:
```py
from qibo.backends import NumpyBackend, TensorflowBackend
from qibojit.backends import NumbaBackend, CupyBackend
```

In the current PSR code, you don't need to worry about this because the backend is specified (either to the global one or something else) when defining the hamiltonian and you can access it via `hamiltonian.backend`. This is because `Hamiltonian` is backend dependent (requires a backend in order to be instantiated), unlike `Circuit` which is backend agnostic (the same circuit object can be executed with different backends).

Reply:
Thank you very much @stavros11 for the explanation. I've modified the code as you suggested. Do you think it is okay now?

Reply:
Yes, it is okay now, thanks.